File: spring_java/src/main/java/dev/pollito/spring_java/config/security/jwt/JwtService.java
package dev.pollito.spring_java.config.security.jwt;

import static io.jsonwebtoken.Jwts.parser;
import static io.jsonwebtoken.security.Keys.hmacShaKeyFor;
import static java.nio.charset.StandardCharsets.UTF_8;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import java.util.Date;
import java.util.function.Function;
import javax.crypto.SecretKey;
import org.jetbrains.annotations.Contract;
import org.jspecify.annotations.NonNull;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

@Service
public class JwtService {

  private final JwtProperties jwtProperties;

  public JwtService(JwtProperties jwtProperties) {
    this.jwtProperties = jwtProperties;
  }

  public String generateToken(@NonNull UserDetails userDetails) {
    return Jwts.builder()
        .subject(userDetails.getUsername())
        .issuedAt(new Date())
        .expiration(new Date(System.currentTimeMillis() + jwtProperties.expirationMs()))
        .signWith(getSignInKey())
        .compact();
  }

  public boolean isTokenValid(String token, @NonNull UserDetails userDetails) {
    return extractUsername(token).equals(userDetails.getUsername()) && !isTokenExpired(token);
  }

  public String extractUsername(String token) {
    return extractClaim(token, Claims::getSubject);
  }

  private boolean isTokenExpired(String token) {
    return extractClaim(token, Claims::getExpiration).before(new Date());
  }

  private <T> T extractClaim(String token, @NonNull Function<Claims, T> claimsResolver) {
    try {
      return claimsResolver.apply(
          parser().verifyWith(getSignInKey()).build().parseSignedClaims(token).getPayload());
    } catch (ExpiredJwtException e) {
      return claimsResolver.apply(e.getClaims());
    }
  }

  @Contract(" -> new")
  private @NonNull SecretKey getSignInKey() {
    return hmacShaKeyFor(jwtProperties.secret().getBytes(UTF_8));
  }
}

File: spring_kotlin/src/main/kotlin/dev/pollito/spring_kotlin/config/security/jwt/JwtService.kt
package dev.pollito.spring_kotlin.config.security.jwt

import io.jsonwebtoken.Claims
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts.builder
import io.jsonwebtoken.Jwts.parser
import io.jsonwebtoken.security.Keys.hmacShaKeyFor
import java.lang.System.currentTimeMillis
import java.nio.charset.StandardCharsets.UTF_8
import java.util.Date
import javax.crypto.SecretKey
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service

@Service
class JwtService(private val jwtProperties: JwtProperties) {

  fun generateToken(userDetails: UserDetails): String {
    return builder()
        .subject(userDetails.username)
        .issuedAt(Date(currentTimeMillis()))
        .expiration(Date(currentTimeMillis() + jwtProperties.expirationMs))
        .signWith(signInKey())
        .compact()
  }

  fun isTokenValid(token: String, userDetails: UserDetails): Boolean {
    return extractUsername(token) == userDetails.username && !isTokenExpired(token)
  }

  fun extractUsername(token: String): String? {
    return extractClaim(token, Claims::getSubject)
  }

  fun <T> extractClaim(token: String, claimsResolver: (Claims) -> T): T? {
    return try {
      claimsResolver(extractAllClaims(token))
    } catch (e: ExpiredJwtException) {
      claimsResolver(e.claims)
    }
  }

  private fun isTokenExpired(token: String): Boolean {
    return extractClaim(token, Claims::getExpiration)?.before(Date()) ?: true
  }

  private fun extractAllClaims(token: String): Claims {
    return parser().verifyWith(signInKey()).build().parseSignedClaims(token).payload
  }

  private fun signInKey(): SecretKey {
    return hmacShaKeyFor(jwtProperties.secret.toByteArray(UTF_8))
  }
}
